core/src/main/java/org/infinispan/transaction/impl/TransactionTable.java

```diff
@@ -508,9 +508,8 @@ protected final void recalculateMinTopologyIdIfNeeded(CacheTransaction removedTr
    public void onTopologyChange(TopologyChangedEvent<?, ?> tce) {
       // don't do anything if this cache is not clustered
       if (clustered) {
-         if (tce.isPre()) {
+         if (!tce.isPre()) {
```

Comment:
Yeah, it is kind of tricky to get the prepare to happen just at the right moment. I started working on it, but I may not finish it today, so I've just added a bit of a comment instead :)